### DISCUSSION TITLE
Sending an effect action in OnInitilized

### DISCUSSION THREAD

**Discussion by janusqa** [mrpmorris/Fluxor#479] [General]
Need some assistance.
I am sending an Effects action in ```OnInitialized```.
The effect calls and api and eventually retrieves that data and sets state.
I've injected the state that is eventually updated on the page, however when that piece of state is eventually updated the page is not re-rendering.
My hope was that when that state changed the page would detect it and re-reneder via the fluxor plumbing.
Is there a pattern for this that works?

**Comment by mfmadsen** [comment]
Please check the documentation at https://github.com/mrpmorris/Fluxor/tree/master/Source/Tutorials/02-Blazor/02A-StateActionsReducersTutorial

The page/component does automatically trigger a re-render on state update if the page/component inherits from FluxorComponent as described in the above docs.

**Comment by janusqa** [comment]
That is the problem I am having, It does not seem to be re-rendering on state changed.
At first I was using this way to hook the page up
```
    // Fluxor Plumbing
    protected override void OnAfterRender(bool firstRender)
    {
        if (firstRender)
        {
            CartState.StateChanged += AppStateHasChanged;
            AuthState.StateChanged += AppStateHasChanged;
        }
    }

    private void AppStateHasChanged(object? sender, EventArgs args)
    {
        InvokeAsync(StateHasChanged);
    }

    public void Dispose()
    {
        CartState.StateChanged -= AppStateHasChanged;
        AuthState.StateChanged -= AppStateHasChanged;
    }
  ```
  Then I tried the other approach....
  ```
  @inherits Fluxor.Blazor.Web.Components.FluxorComponent
  ```
  Neither causes the page to re-render when state is changed due to calling that async Effect action.  
  I can see that the state has changed in Redux toolkit.
  
  https://learn.microsoft.com/en-us/aspnet/core/blazor/components/lifecycle?view=aspnetcore-8.0
  I thought maybe the section ```Handle incomplete async actions at render``` maybe related, but came to the conclusion that fluxor handles the state change notifications in this case.
  
  I am wondering if it is because its an action to and async Effect, but do not see why that would be an issue.

**Comment by mrpmorris** [comment]
Can you create a minimal repro?

**Comment by janusqa** [comment]
@mrpmorris 
This is a link to a Repro Repo.  Imagine, if you would, that the table is a form and that forecasts is a model for the form.
https://github.com/janusqa/blazorbugs
Experimental page: https://localhost:7079/weather

P.S. :I have been experimenting over the last days, and realize that it is what I want to do that is at issue here, and that I am in need of a pattern to solve it.
The context is that I have a form and a model I want to bind to that form. 

Some context:
1. I want to dispatch an ```Effect``` action to retrieve data from a server
2. Update the state related to that information which is done by the ```Action```, 
3. copy the state information to the model 
4. display that information in the form via the model (This in conjunction with ```3``` above is what is the root of the problem)

If I refresh (or visit for the first time) the page I had wanted to see the information in the form, and likewise, if I navigate to and back.
The problem happens when I go to the page for the first time. The information is fetched into state but is not displayed in the form unless i navigate away and back.

As it is now if you refresh the page, it is stuck on loading.  State due to being fetched is updating too late to be put into forecasts.
I was initially setting forecast in ```OnInitializedAsync``` to handle the navigate away and back, and in the ```OnAfterRender``` (to handle the refresh) but of course it would not be displayed as the page was already rendered.

There is what I call a hacky fix to get the desired result where forecasts is set only in ```OnAfterRender``` and only after state is updated from the fetch and only on the first time it was set (to prevent a re-render infinite loop).
